app/src/main/java/com/nextcloud/talk/translate/TranslateActivity.kt

```diff
@@ -0,0 +1,276 @@
+        binding = ActivityTranslateBinding.inflate(layoutInflater)
+
+        setupActionBar()
+        setContentView(binding.root)
+        setupSystemColors()
+        setupTextViews()
+        setupSpinners()
+        getLanguageOptions()
+        translate(null, Locale.getDefault().language)
+    }
+
+    private fun setupActionBar() {
+        setSupportActionBar(binding.translationToolbar)
+        binding.translationToolbar.setNavigationOnClickListener {
+            onBackPressed()
+        }
+        supportActionBar?.setDisplayHomeAsUpEnabled(true)
+        supportActionBar?.setDisplayShowHomeEnabled(true)
+        supportActionBar?.setIcon(ColorDrawable(resources!!.getColor(R.color.transparent)))
+        supportActionBar?.title = resources!!.getString(R.string.translation)
+        viewThemeUtils.material.themeToolbar(binding.translationToolbar)
+    }
+
+    private fun setupTextViews() {
+        val original = binding.originalMessageTextview
+        val translation = binding.translatedMessageTextview
+
+        viewThemeUtils.talk.themeIncomingMessageBubble(original, grouped = true, deleted = false)
+        viewThemeUtils.talk.themeIncomingMessageBubble(translation, grouped = true, deleted = false)
+
+        original.movementMethod = ScrollingMovementMethod()
+        translation.movementMethod = ScrollingMovementMethod()
+
+        val bundle = intent.extras
+        binding.originalMessageTextview.text = bundle?.getString(BundleKeys.KEY_TRANSLATE_MESSAGE)
+        text = bundle?.getString(BundleKeys.KEY_TRANSLATE_MESSAGE)
+    }
+
+    private fun getLanguageOptions() {
+        val currentUser: User = userManager.currentUser.blockingGet()
+        val json = JSONArray(CapabilitiesUtilNew.getLanguages(currentUser).toString())
+        Log.i(TAG, "json is: $json")
+
+        val fromLanguagesSet = mutableSetOf(resources.getString(R.string.translation_detect_language))
+        val toLanguagesSet = mutableSetOf(resources.getString(R.string.translation_device_settings))
+
+        for (i in 0..json.length() - 1) {
+            val current = json.getJSONObject(i)
+            if (current.getString(FROM_ID) != Locale.getDefault().language) {
+                toLanguagesSet.add(current.getString(FROM_LABEL))
+            }
+
+            fromLanguagesSet.add(current.getString(TO_LABEL))
+        }
+
+        fromLanguages = fromLanguagesSet.toTypedArray()
+        toLanguages = toLanguagesSet.toTypedArray()
+
+        fillSpinners()
+    }
+
+    private fun enableSpinners(value: Boolean) {
+        binding.fromLanguageSpinner.isEnabled = value
+        binding.toLanguageSpinner.isEnabled = value
+    }
+
+    private fun translate(fromLanguage: String?, toLanguage: String) {
+        val currentUser: User = userManager.currentUser.blockingGet()
+        val credentials: String = ApiUtils.getCredentials(currentUser.username, currentUser.token)
+        val translateURL = currentUser.baseUrl +
```

Comment:
lets extract the url creation to ApiUtils where all url's are built...

Reply:
fixed via https://github.com/nextcloud/talk-android/pull/3025/commits/b779a01c274919c464a40ec3e71cfc947071bb92